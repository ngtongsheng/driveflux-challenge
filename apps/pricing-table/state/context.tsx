import React, { createContext, useReducer, Dispatch } from 'react';
import { Pricing } from '@driveflux-code-challenge/interfaces';

export const START_EDIT = 'START_EDIT';
export const CANCEL_EDIT = 'CANCEL_EDIT';
export const UPDATE_PRICING_FIELD = 'UPDATE_PRICING_FIELD';
export const RESET_ALL_PRICINGS = 'RESET_ALL_PRICINGS';
export const START_SAVE_PRICINGS = 'START_SAVE_PRICINGS';
export const SUCCESS_GET_PRICINGS = 'SUCCESS_GET_PRICINGS';
export const SUCCESS_SAVE_PRICINGS = 'SUCCESS_SAVE_PRICINGS';

interface Action {
  type: string;
  payload?: PricingField | Pricings;
}

interface PricingField {
  id: string;
  field: string;
  value: string;
}

interface Pricings {
  snapshot?: Pricing[];
  pricings: Pricing[];
}

interface PricingState extends Pricings {
  readOnly: boolean;
  isSaving: boolean;
}

const initialState: PricingState = {
  snapshot: [],
  pricings: [],
  readOnly: true,
  isSaving: false
};

const reducer = (state: PricingState, { type, payload }: Action): PricingState => {
  switch (type) {
    case SUCCESS_GET_PRICINGS: {
      const { pricings } = payload as Pricings;
      return {
        ...state,
        pricings
      };
    }

    case START_EDIT: {
      const { pricings } = state;
      return {
        ...state,
        snapshot: [...pricings],
        readOnly: false
      };
    }

    case CANCEL_EDIT: {
      const { snapshot } = state;
      return {
        ...state,
        snapshot: [],
        pricings: [...snapshot],
        readOnly: true
      };
    }

    case UPDATE_PRICING_FIELD: {
      const { id, field, value } = payload as PricingField;
      const newValue = Math.max(0, parseFloat(value));

      const pricings = state.pricings.map(pricing => {
        if (pricing.id !== id) {
          return pricing;
        }

        if (field !== "lite") {
          return {
            ...pricing,
            [field]: String(newValue)
          }
        }

        return {
          ...pricing,
          [field]: String(newValue),
          standard: String(newValue * 2),
          unlimited: String(newValue * 3),
        }
      })

      return {
        ...state,
        pricings
      };
    }

    case RESET_ALL_PRICINGS: {
      const pricings = state.pricings.map(pricing => ({
        ...pricing,
        lite: "0",
        standard: "0",
        unlimited: "0"
      }))

      return {
        ...state,
        pricings
      };
    }

    case START_SAVE_PRICINGS: {
      return {
        ...state,
        isSaving: true
      };
    }

    case SUCCESS_SAVE_PRICINGS: {
      return {
        ...state,
        isSaving: false,
        readOnly: true,
        snapshot: []
      };
    }

    default:
      return state;
  }
};

export const PricingContext = createContext<[s: PricingState, d: Dispatch<Action>]>(
  [null, null]
);

export const PricingContextProvider = ({ children }) => {
  const [state, dispatch] = useReducer(reducer, initialState);

  return (
    <PricingContext.Provider value={[state, dispatch]}>
      {children}
    </PricingContext.Provider>
  );
};
