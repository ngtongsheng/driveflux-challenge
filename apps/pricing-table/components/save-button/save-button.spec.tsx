import React, { useContext } from 'react';
import axios from 'axios';
import { render, cleanup, fireEvent, waitFor } from '@testing-library/react';

import { PricingContext, PricingContextProvider } from '../../state/context';
import SaveButton from './save-button';

jest.mock('axios');
const mockedAxios = axios as jest.Mocked<typeof axios>;

const setup = () => {
  const Component = () => {
    const [state] = useContext(PricingContext);
    const { snapshot, readOnly, isSaving, pricings } = state;

    return (
      <div>
        <SaveButton />
        <div className="state">
          <span id="isSaving">{String(isSaving)}</span>
          <span id="readOnly">{String(readOnly)}</span>
          <span id="snapshot">
            {Object.values(snapshot[0] || {}).join(',')}
          </span>
          <span id="pricings">
            {Object.values(pricings[0] || {}).join(',')}
          </span>
        </div>
      </div>
    );
  };

  const utils = render(
    <PricingContextProvider
      overrideState={{
        snapshot: [
          {
            id: 'test',
            label: 'test',
            lite: '0',
            standard: '0',
            unlimited: '0',
          },
        ],
        pricings: [
          {
            id: 'test',
            label: 'test',
            lite: '2',
            standard: '4',
            unlimited: '6',
          },
        ],
        readOnly: false,
      }}
    >
      <Component />
    </PricingContextProvider>
  );
  const { container } = utils;
  const button = container.querySelector('.button');
  const state = container.querySelector('.state');

  return {
    ...utils,
    button,
    state,
  };
};

afterEach(cleanup);

describe('SaveButton', () => {
  it('should render successfully', () => {
    const { baseElement, button } = setup();
    expect(baseElement).toBeTruthy();
    expect(button).toMatchSnapshot();
  });

  it('should be successfully saved', async () => {
    const { button, state } = setup();

    mockedAxios.put.mockResolvedValue({ success: true });

    expect(state.querySelector('#readOnly').innerHTML).toEqual('false');
    expect(state.querySelector('#isSaving').innerHTML).toEqual('false');
    expect(state.querySelector('#snapshot').innerHTML).toEqual(
      'test,test,0,0,0'
    );

    fireEvent.click(button);

    expect(state.querySelector('#isSaving').innerHTML).toEqual('true');

    await waitFor(() => {
      expect(state.querySelector('#isSaving').innerHTML).toEqual('false');
      expect(state.querySelector('#readOnly').innerHTML).toEqual('true');
      expect(state.querySelector('#snapshot').innerHTML).toEqual('');
    });
  });

  it('should be reset edit mode if error', async () => {
    const { button, state } = setup();

    mockedAxios.put.mockRejectedValue({});

    expect(state.querySelector('#readOnly').innerHTML).toEqual('false');
    expect(state.querySelector('#isSaving').innerHTML).toEqual('false');
    expect(state.querySelector('#snapshot').innerHTML).toEqual(
      'test,test,0,0,0'
    );

    fireEvent.click(button);

    expect(state.querySelector('#isSaving').innerHTML).toEqual('true');

    await waitFor(() => {
      expect(state.querySelector('#isSaving').innerHTML).toEqual('false');
      expect(state.querySelector('#readOnly').innerHTML).toEqual('true');
      expect(state.querySelector('#snapshot').innerHTML).toEqual('');
      expect(state.querySelector('#pricings').innerHTML).toEqual(
        'test,test,0,0,0'
      );
    });
  });
});
