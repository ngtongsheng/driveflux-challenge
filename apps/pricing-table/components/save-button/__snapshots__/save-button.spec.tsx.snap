// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SaveButton should render successfully 1`] = `
<button
  class="jsx-2282515019 button is-dark"
>
  Save
</button>
`;
